rust/worker/src/segment/distributed_hnsw_segment.rs

```diff
@@ -207,40 +200,16 @@ impl<'a> SegmentWriter<'a> for DistributedHNSWSegmentWriter {
 
                     index.add(record.offset_id as usize, embedding);
                 }
-                // This shouldn't be reached since materialization always derefs
-                // upserts into either updates or inserts.
-                Operation::Upsert => {
-                    panic!(
-                        "Invariant violation. Upserts should not be present after materialization"
-                    );
-                }
-                Operation::Update => {
-                    // Should panic here if embedding is not found because it likely
-                    // means that somehow our storage is corrupt as data record on
-                    // the record segment does not contain the embedding.
-                    let embedding = match record.final_embedding {
-                        Some(e) => e,
-                        None => match record.data_record.as_ref() {
-                            Some(record) => record.embedding,
-                            None => {
-                                panic!("Invariant violation. Embedding not found on storage");
-                            }
-                        },
-                    };
-                    // HNSW index behavior is to treat add() as upsert so this
-                    // will update the embedding if it exists. It does not
-                    // perform any validation on its own and assumes that the
-                    // offset ids are correct (i.e. pertaining to records that
-                    // are actually meant to be updated).
-                    self.index.read().add(record.offset_id as usize, embedding);
-                }
-                Operation::Delete => {
+                MaterializedLogOperation::DeleteExisting => {
                     // HNSW segment does not perform validation of any sort. So,
                     // the assumption here is that the materialized log records
                     // contain the correct offset ids pertaining to records that
                     // are actually meant to be deleted.
                     self.index.read().delete(record.offset_id as usize);
                 }
+                MaterializedLogOperation::Initial => panic!(
```

Comment:
would be nice if this was not possible, initial feels internal to materialization